src/error.rs

```diff
@@ -1,25 +1,111 @@
-use crate::response::Response;
-use std::fmt;
-use std::io::{self, ErrorKind};
+use url::Url;
+
+use std::error;
+use std::fmt::{self, Display};
+use std::io::{self};
+
+use crate::Response;
 
 #[derive(Debug)]
-pub enum Error {
+pub struct Error {
+    kind: ErrorKind,
+    message: String,
+    url: Option<Url>,
+    source: Option<Box<dyn error::Error>>,
+    response: Option<Box<Response>>,
+}
+
+impl Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        if let Some(url) = &self.url {
+            write!(f, "{}: ", url)?;
+        }
+        if let Some(response) = &self.response {
+            write!(f, "status code {}", response.status())?;
+        } else {
+            write!(f, "{:?}: {}", self.kind, self.message)?;
+        }
+        if let Some(source) = &self.source {
+            write!(f, ": {}", source)?;
+        }
+        Ok(())
+    }
+}
+
+impl error::Error for Error {
+    fn source(&self) -> Option<&(dyn error::Error + 'static)> {
+        self.source.as_deref()
+    }
+}
+
+impl Error {
+    pub(crate) fn new(kind: ErrorKind, message: String) -> Self {
+        Error {
+            kind,
+            message,
+            url: None,
+            source: None,
+            response: None,
+        }
+    }
+
+    pub(crate) fn url(mut self, url: Url) -> Self {
```

Comment:
In two minds about exposing `Url` as a type here. Since our entry point is mainly `&str` 🤔

Reply:
In this case it's only exposed at the crate level, so doesn't affect the public-facing API; we can always change it later.